[core,tls] Use BIO_wait_read replace event

KangLin opened the pull request:
Events may be invalid, so you should use BIO_wait_read.

ag:
QTcpSocket, If the socket is using [QNetworkProxy](https://doc.qt.io/qt-6/qnetworkproxy.html), the returned descriptor may not be usable with native socket functions.
See: https://doc.qt.io/qt-6/qabstractsocket.html#socketDescriptor

KangLin committed:
[core,tls] Use BIO_wait_read replace event

akallabeth reviewed (changes requested):
Check `transport_layer_bio_ctrl` for an implementation of this.
your change breaks the transport abstraction (wait for stuff other than the socket), sorry :/

akallabeth commented:
@KangLin Consider the direct connection (no gateway):

* `BIO_wait_read` will eventually call `freerdp_tcp_layer_wait` and block there in `poll`. no way to wait for abortEvent et al.
* `freerdp_tcp_layer_get_event` will wait for `freerdp_tcp_connect_layer` to set the `WSAEvent` (which wraps native sockets) and can be used to wait for multiple handles

akallabeth commented:
so, this sounds like you need your own implementation of that layer based on `QTcpSocket`?

KangLin commented:
> @KangLin do I see this correct that you try to open a `socket` and then pass that to `FreeRDP`?
> 
> that is already available with default implementation, see `freerdp_tcp_default_connect` [edit] documentation is not good for this, so here is an example: `/v:|:1234` will directly use the socket `1234`

@akallabeth 
- There is no problem using pipes and unix sockets. [CChannelSSHTunnelForward](https://github.com/KangLin/RabbitRemoteControl/blob/master/Client/SSH/ChannelSSHTunnelForward.h)
- I implemented a connection layer based on libssh: [ConnectLayerSSHTunnel](https://github.com/KangLin/RabbitRemoteControl/blob/master/Plugins/FreeRDP/Client/ConnectLayerSSHTunnel.h). It has a nation socket. so using event and wait read is the same.
- I implemented a connection layer based on QTcpSocket:  [CConnectLayerQTcpSocket](https://github.com/KangLin/RabbitRemoteControl/blob/master/Plugins/FreeRDP/Client/ConnectLayerQTcpSocket.h). However, QTcpSocket may not be converted to event, so in this case you can only use wait read.

KangLin commented:
@akallabeth 
> in any case, if you have a problem in your implementation, then I'd check that first, as mentioned before, the current implementation can already reuse an already opened socket without any modifications.

There is nothing problem with my current implementation.

> We're trying to remove all locations where you need a blocking (unconditional) wait that can not be broken by freerdp_abort_connect_context.
>
> Here we have a blocking loop with a timeout.
the timeout should be eliminated and a infinite wait used instead (with the aforementioned abort condition)
Only if that event is triggered the read should be done, not before.

#11651 I don't think this solves the problem. 

I think pollAndHandshake() and similar blocking functions should not be provided.

I think the final approach is the design should be based on event-driven. No need to block internally.

When data cannot be read, it returns directly to the main event loop and waits for the next time data can be read, and then executes according to the current status again.

The current BIO design has two layers, the upper layer is the memory cache and the lower layer is the connection layer.
When the data to be read is larger than the data in the cache, you can directly exit(don't need blocking) and enter the main event loop.
